### ISSUE TITLE
Problem with nuxt imports: Vitest Coverage stays at 100% for components regardless of the existence of any component test because of nuxt imports

### ISSUE THREAD

**Issue by FlorianPhilipp2007** [nuxt/test-utils#891]
### Environment

- Operating System: Darwin
- Node Version:     v18.16.0
- Nuxt Version:     3.11.2
- CLI Version:      3.11.1
- Nitro Version:    2.9.6
- Package Manager:  npm@9.5.1
- Builder:          -
- User Config:      devtools, typescript, extends, components, ssr, runtimeConfig, app, css, build, modules, experimental, i18n, proxy, vite, public, nitro, hooks
- Runtime Modules:  @pinia/nuxt@0.5.1, nuxt-proxy@0.4.1, vuetify-nuxt-module@0.13.5, @nuxt/test-utils/module@3.13.1, @nuxtjs/i18n@8.3.1
- Build Modules:    -

### Reproduction

https://stackblitz.com/edit/nuxt-starter-gvfsue?file=nuxt.config.ts,components%2Fnested%2FTestNestedUntested.vue

vitest --coverage is not working in StackBlitz. Please download the project and try it locally.

### Describe the bug

Description:
When the components option in the Nuxt configuration is set to true (enabling automatic global registration of components), the test coverage report shows 100% coverage for all components, regardless of whether tests actually exist for them. Something is importing the untested TestNestedUntested. V8 is reporting coverage for it. Maybe Nuxt imports all those files. 

Actual Behavior:
The coverage report incorrectly shows 100% coverage for all components, including those without any tests like TestNestedUntested.vue in component directory. Seems to be a problem caused by nuxt importing all files. 

Expected Behavior:
The coverage report should accurately reflect the coverage for each component. Only TestNested.vue should show 100% coverage, while TestNestedUnested.vue should show 0% coverage or appropriate coverage based on the absence of tests.

**Comment by FlorianPhilipp2007** [comment]
> This is unlikely something to fix in Nuxt. It might possibly be a vitest bug or something to be resolved in nuxt/test-utils. Have you tried ensuring that `.nuxt` is added to your coverage exclude patterns?

I first raised the issue at vitest but they think its a nuxt problem: 
https://github.com/vitest-dev/vitest/issues/6029#issuecomment-2210914420

**Comment by danielroe** [comment]
It seems like the issue is the one Ari linked - it's a Vue sourcemap bug: https://github.com/vitest-dev/vitest/issues/3607#issuecomment-1608954728.

**Comment by AriPerkkio** [comment]
This report contains two issues:
1. Something (nuxt ?) is importing files that are not being tested. Node's V8 coverage includes every file that is loaded via Node. The repro is testing `TestNested.vue`, but is also unexpectedly seeing coverage for `UntestNested.vue`. Vitest is not importing `UntestNested.vue`. 
2. Coverage is incorrectly 100%. This is caused by Vue's sourcemaps being off. We've seen so many of these issues that I think I'll need to raise an issue on Vue some day.

**Comment by DenisLug** [comment]
@AriPerkkio, @danielroe: Thank you both for your help so far.

The issue with the coverage report has evolved after updating vitest to version 2 ([see comment in vitest issue](https://github.com/vitest-dev/vitest/issues/6029#issuecomment-2296683167), [Stackblitz demo](https://stackblitz.com/edit/nuxt-starter-q9e4id)). Previously, the coverage incorrectly showed 100% for all components, but now it shows 0% coverage for files that do have tests.

To summarize the current status:

- With the `istanbul` provider, the coverage report collects only the coverage for the utils directory
- With the `v8` provider, the coverage report now shows 0% coverage for files that have tests

Given these changes, could we clarify who will be investigating this further? Is this something that should be addressed by the Nuxt team, or should we continue the discussion with the Vitest team?

**Comment by DenisLug** [comment]
I have retested the issue with the new vitest major version 3 and the issue still persists. See here for the updated reproduction: https://stackblitz.com/edit/github-hfxk3ssz

**Comment by jsedanoj** [comment]
https://github.com/nuxt/nuxt/issues/31384 has been closed as duplicate of this one, so I will share the diagnosis here (see the link for more detail, instructions, screenshots and reproduction in a repo, including the workaround).

Starting from the basic Nuxt application as bootstrapped from the documentation, when I add .vue files to the pages/ directory, the coverage is not measured. The other folders work. Renaming the directory to pagesX/ solves the coverage, but breaks the application (obviously). Configuring nuxt.config:dirs.pages to pagesX fixes the application, but breaks the coverage again.

Apparently, Nuxt is doing something with the pages/ directory that breaks the coverage.

**Comment by danielroe** [comment]
@jsedanoj if you think it's a different issue we can open a new one but it should be in https://github.com/nuxt/test-utils.

**Comment by jsedanoj** [comment]
Well, you know better than me. If you think the underlying problem is the same, we can keep it here. If you prefer to create a new one for tracking (because the symptoms are different) I can create it, no problem. Just tell me and I will (I have the repo with the reproduction and the text, so it is a matter of minutes). I'll do as you prefer.

**Comment by MatteoGabriele** [comment]
I'm experiencing a similar issue. I created a small application to identify which package was causing the problem, and it appears to be Vitest v3 when testing a Nuxt 3 project. When I run the same code and tests using Vitest v2, the coverage report is accurate. However, with v3, the components inside the /pages directory are ignored. I’ve attached a simple side-by-side screenshot comparing Vitest v2 and v3. Let me know if you need a reproduction for further investigation; I’d be happy to help.

![Image](https://github.com/user-attachments/assets/2b51df2f-01c5-488d-baf2-3f65956c8172)

**Comment by jsedanoj** [comment]
If you already have the project done, maybe you could push it to a repo, as example of workaround.
If it works for me, maybe I can just downgrade my vitest, instead of using my workaround.

This may make us think that the problem is inside vitest, actually. But considering that it only fails for one directory of nuxt... and being this directory "pages/" that we know it has special meaning for nuxt...

**Comment by MatteoGabriele** [comment]
This repository contains a component and a page with the same code and tests. The main branch uses Vitest 3 but with the broken code coverage. There is another branch called "vitest2" that has working code coverage, which involves downgrading to Vitest v2 and @vitest/coverage-v8 v2.

I'm uncertain about which library is causing the issue. It appears to be related to Vitest, but since the problem involves components within the pages folder, I suspect that Nuxt might be encountering an issue with something in Vitest v3.

**Comment by ademyalcin27** [comment]
I have an example here , I commented some of test, it show 100%. 
https://stackblitz.com/edit/nuxt-starter-gjkmvknv?file=pages%2Findex.vue

<img width="699" alt="Image" src="https://github.com/user-attachments/assets/ca3ed978-2742-4bfa-bf49-7b5dd40de4cf" />

Another interesting thing is I have the same configuration in my local but I got different results as below. and sometimes it goes endless loop, so the test is rerun until I stop

<img width="793" alt="Image" src="https://github.com/user-attachments/assets/ea2e7abc-9df7-4171-840e-761ad3ec7a65" />

**Comment by MatteoGabriele** [comment]
@danielroe just wanted to point out that similar issues with code coverage and Vitest v3 have been spotted in the i18n module https://github.com/nuxt-modules/i18n/issues/3238#issuecomment-2672492536 
Maybe helps, maybe it doesn't.

**Comment by jsedanoj** [comment]
And now it is happening for me also I other folders on the project (.vue files in components/). So far, I have found no workaround, but it is systematic.
I will try to add this to my minimal example, if I can simplify it.

In my search through the internet some people says that maybe the sourcemap is incorrect... well, it makes sense. But I do not have enough experience with the sourcemap to say if it is correct or not.

**Comment by MatteoGabriele** [comment]
To eliminate any uncertainties, I have made the exact same application using plain Vue, VueRouter, and Vitest, and it works as expected without any issues. I had to try it.

**Comment by MatteoGabriele** [comment]
The coverage will work if I change the folder name of "pages" to "views", for example.
It's important to note that if I change the value of `dir.pages` from the default setting of "pages" to something like "views," the coverage will fail again.

<img width="927" alt="Image" src="https://github.com/user-attachments/assets/aab35ee6-7ce3-4ded-a2a9-25c122083ec0" />

**Comment by ineshbose** [comment]
From my research, seems that the thread to follow is: https://github.com/vitest-dev/vitest/issues/4993

**Comment by jsedanoj** [comment]
@MatteoGabriele , did you create a bug report in vitest for this?

**Comment by MatteoGabriele** [comment]
@jsedanoj not really. I thought that this was a nuxt error since it is so specific about pages.

**Comment by MatteoGabriele** [comment]
I completely get the frustration, but just a reminder that open source relies on community effort. It’s important to stay constructive and supportive so we can all work together to improve things.

If your previous setup worked better, it might be worth sticking with that for now. Personally, I'm using the latest Nuxt 3 with Vitest 2 and haven’t run into issues. You can also check my [reproduction](https://github.com/MatteoGabriele/nuxt-vitest-project) where the "vitest2" branch works as expected.

Most importantly, let’s not forget to appreciate the hard work that goes into these projects. Gratitude really goes a long way.

**Comment by jsedanoj** [comment]
Yes, you are right, the frustration was speaking.

The problem for one of those "previous setups" is that... it was Nuxt2, out of support now. So not an option anymore.

**Comment by MatteoGabriele** [comment]
@jsedanoj, no worries! 
If you are using Nuxt 2, you must have had a moment before the recent updates when you were able to achieve the correct coverage, right?
Have you tried using an exact version of Vitest 2? It may be a tedious job, but consider reverting to several earlier minor versions until you find one that works.

**Comment by jsedanoj** [comment]
Not really.

In one of the projects, I was migrating from Nuxt2 to Nuxt3, and I thought it worked because my tests where giving 100% coverage... until I have discovered, during this investigation, that it actually didn't, it just always says 100% with v8, and completely non-sensical lines with istambul.

In another project, I was moving from another architecture to Nuxt3 directly... so yes, I will just stay in the other architecture (express + pure Vue3), it works and is well maintained.

So it is not because of lack of options.

**Comment by AriPerkkio** [comment]
I have 0 experience of Nuxt, but the minimal reproduction seems ot work when setting `components.global` to `false` in `nuxt.config.ts`.

<img src="https://github.com/user-attachments/assets/4f66301d-09a9-4c00-8dbc-644e8ec26ba3" width="400" />

When it's set to `true`, coverage report is odd:

<img src="https://github.com/user-attachments/assets/a8e16a7c-452c-4688-b775-95840fc4abeb" width="400" />

What exactly does `components.global` do? How should testing tools handle it?

**Comment by danielroe** [comment]
generally speaking people _should not_ set `components.global`. It registers a dynamic import of every single component with vue. But it's sometimes used by people who want components to be available because they don't know what will be used (e.g. they are getting the data from a CMS)...

Properly speaking this shouldn't increase the coverage because the component isn't _run_.

**Comment by jsedanoj** [comment]
Sorry, @danielroe , do you mean that we should not configure "components.global:true2? Or do you mean that we should not change the default (which seems to be "true")?

**Comment by danielroe** [comment]
@jsedanoj The default is `false`, and you should not configure it unless you have a specific reason to do so. It will produce lots of async chunks (one for every component) for your site and may impact your site's performance.

**Comment by arkoannuk** [comment]
Running into the same problem, where coverage about `/components` isn't correctly reported.

Upgrading to `@vitejs/plugin-vue" "^5.2.4` did not help, like suggested [here](https://github.com/vuejs/core/issues/13260#issuecomment-2866643890), although I'm not sure if that issue is directly related to this.

**Comment by Serhansolo** [comment]
Guys, I have the same problem, and I am not using Nuxt, just Vue 3, Pinia and the Router. Testing my components mostly with testing library. 



I see a red line with 0% coverage when: 
* When I don't import the file anywhere.
* Whenever I import a file in a random test file (*.spec.ts), it does not matter if I do or do not include a test where I render the component. 
* Whenever I import the component inside another (tested) component but not use it in the template. I see a red line telling me I have 0% coverage.


I see 100% coverage when:
* I import the component inside another (tested) component and use it in the template. While not testing/rendering the component anywhere. So just the import only.

Gray area:
* Probably also if I would make tests to actually cover statements and branches, but I would not know since I get 100% coverage as soon as I import the component.


Hopefully I can do a bit more investigation, because this does not make any sense. I should, in my humble opinion, just be able to trust the cov-report.
